### PR TITLE
Performance guidelines

### LINKED ISSUE #212: Performance Guidance
Some things already mentioned in various docs. Would be nice to have one doc dedicated to all the perf tips.

Some ideas:
- [ ] Streaming vs buffering (IEnumerable/IAsyncEnumerable vs. ToList/ToListAsync). Possibly refer from https://docs.microsoft.com/en-us/ef/core/querying/how-query-works#when-queries-are-executed.
- [ ] Split/single queries
- [ ] Perf counters
- [ ] Tracking vs no-tracking (per query, per type, etc.)
- [ ] DbContext pooling
- [ ] Compiled queries
- [ ] FromSql for complex queries
- [ ] INotifyPropertyChanged
- [ ] Key generation: sequence vs identity
- [ ] Batching (configuring sizes, networking conditions, etc)

### LINKED ISSUE #2608: Note potential performance implications
It should be noted in documentation that using async operations in EF Core backed by sqlclient may result in very severe performance degradation over a synchronous option.

https://github.com/dotnet/efcore/issues/22331

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 763dd10a-6619-d9b0-28fb-d9a0be0048b5
* Version Independent ID: 81b05619-22c4-cd7c-7d27-fcb08b888a77
* Content: [Asynchronous Queries - EF Core](https://docs.microsoft.com/en-us/ef/core/querying/async)
* Content Source: [entity-framework/core/querying/async.md](https://github.com/dotnet/EntityFramework.Docs/blob/master/entity-framework/core/querying/async.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @smitpatel
* Microsoft Alias: **avickers**

### PR DESCRIPTION
**Pull request by roji**
Here's a draft for a new performance guidelines section, any and all feedback is welcome.

Note that code is currently inline, I will separate it out into the samples before merging.

Closes #212
Closes #2608

### CODE REVIEW

File: entity-framework/core/miscellaneous/async.md [hunk 1]
```diff
@@ -18,6 +18,9 @@ For more information, see [the general C# asynchronous programming docs](/dotnet
 > [!WARNING]
 > EF Core doesn't support multiple parallel operations being run on the same context instance. You should always wait for an operation to complete before beginning the next operation. This is typically done by using the `await` keyword on each async operation.
 
+> [!WARNING]
+> The async implementation of [Microsoft.Data.SqlClient](https://github.com/dotnet/SqlClient) unfortunately has some known issues with async on .NET Core (e.g. [#593](https://github.com/dotnet/SqlClient/issues/593), [#601](https://github.com/dotnet/SqlClient/issues/601), and others).
```
Comment (ErikEJ) [inline entity-framework/core/miscellaneous/async.md:22]: Could "on .NET Core" be removed

File: entity-framework/core/miscellaneous/connection-resiliency.md [hunk 2]
```diff
@@ -28,6 +28,9 @@ public void ConfigureServices(IServiceCollection services)
 }
 ```
 
+> [!NOTE]
+> Enabling retry on failure causes EF to internally buffer the resultset, which may significantly increase memory requirements for queries returning many results. See [buffering and streaming](core/miscellaneous/performance#buffering-and-streaming) for more details.
```
Comment (ErikEJ) [inline entity-framework/core/miscellaneous/connection-resiliency.md:32]: "returning large result sets" ?

File: entity-framework/core/performance/advanced-performance-topics.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+author: rick-anderson
+ms.author: riande
+ms.date: 12/9/2020
+uid: core/performance/advanced-performance-topics
+---
+# Advanced Performance Topics
+
+## DbContext pooling
+
+`AddDbContextPool` enables pooling of `DbContext` instances. Context pooling can increase throughput in high-scale scenarios such as web servers by reusing context instances, rather than creating new instances for each request.
```
Comment (ErikEJ) [inline entity-framework/core/performance/advanced-performance-topics.md:13]: Maybe move the last sentence in this topic up here?

**Comment by ErikEJ** [comment]
Revival of the EF6 performance white paper ? 🥳

File: entity-framework/core/miscellaneous/async.md (hunk 1 above)
Comment (roji) [reply entity-framework/core/miscellaneous/async.md:22]: Yeah, missed that it was an issue in netfx too, will do.

**Comment by roji** [comment]
> Revival of the EF6 performance white paper ?

Yep :)

File: entity-framework/core/performance/advanced-performance-topics.md (hunk 3 above)
Comment (roji) [reply entity-framework/core/performance/advanced-performance-topics.md:13]: Are you referring to the paragraph that starts with "Context pooling is intended for scenarios where the context configuration, which includes services resolved, is fixed between requests."? It seems OK to me under the limitations, what do you have in mind?

**Review by bricelam — APPROVED** [review]
🐑🇮🇹

Loved this. So much good information with a clean layout. You may want to run [a grammar and spell check](https://microsoftedge.microsoft.com/addons/detail/microsoft-editor-spellin/hokifickgkhplphjiodbggjmoafhignh) before merging, since I didn't bother marking the ones I saw while reading it. (Since I read the rendered preview on my phone.)

**Review by AndriySvyryd — APPROVED** [review]

**Comment by roji** [comment]
Thanks for the reviews and feedback :)

**Pull request merged**